xds/src/main/java/io/grpc/xds/AbstractXdsClient.java

```diff
@@ -0,0 +1,760 @@
+     * {@code errorDetail}. Used for reacting to a specific discovery response. For
+     * client-initiated discovery requests, use {@link
+     * #sendDiscoveryRequest(ResourceType, Collection)}.
+     */
+    abstract void sendDiscoveryRequest(ResourceType type, String versionInfo,
+        Collection<String> resources, String nonce, @Nullable String errorDetail);
+
+    /**
+     * Sends a client-initiated discovery request.
+     */
+    final void sendDiscoveryRequest(ResourceType type, Collection<String> resources) {
+      String nonce;
+      switch (type) {
+        case LDS:
+          nonce = ldsRespNonce;
+          break;
+        case RDS:
+          nonce = rdsRespNonce;
+          break;
+        case CDS:
+          nonce = cdsRespNonce;
+          break;
+        case EDS:
+          nonce = edsRespNonce;
+          break;
+        case UNKNOWN:
+        default:
+          throw new AssertionError("Unknown resource type: " + type);
+      }
+      sendDiscoveryRequest(type, getCurrentVersion(type), resources, nonce, null);
+    }
+
+    final void handleRpcResponse(
+        ResourceType type, String versionInfo, List<Any> resources, String nonce) {
+      if (closed) {
+        return;
+      }
+      responseReceived = true;
+      // Nonce in each response is echoed back in the following ACK/NACK request. It is
+      // used for management server to identify which response the client is ACKing/NACking.
+      // To avoid confusion, client-initiated requests will always use the nonce in
+      // most recently received responses of each resource type.
+      switch (type) {
+        case LDS:
+          ldsRespNonce = nonce;
+          handleLdsResponse(versionInfo, resources, nonce);
+          break;
+        case RDS:
+          rdsRespNonce = nonce;
+          handleRdsResponse(versionInfo, resources, nonce);
+          break;
+        case CDS:
+          cdsRespNonce = nonce;
+          handleCdsResponse(versionInfo, resources, nonce);
+          break;
+        case EDS:
+          edsRespNonce = nonce;
+          handleEdsResponse(versionInfo, resources, nonce);
+          break;
+        case UNKNOWN:
+        default:
+          logger.log(XdsLogLevel.WARNING, "Ignore an unknown type of DiscoveryResponse");
+      }
+    }
+
+    final void handleRpcError(Throwable t) {
+      handleRpcStreamClosed(Status.fromThrowable(t));
+    }
+
+    final void handleRpcCompleted() {
+      handleRpcStreamClosed(Status.UNAVAILABLE.withDescription("Closed by server"));
+    }
+
+    private void handleRpcStreamClosed(Status error) {
+      checkArgument(!error.isOk(), "unexpected OK status");
+      if (closed) {
+        return;
+      }
+      logger.log(
+          XdsLogLevel.ERROR,
+          "ADS stream closed with status {0}: {1}. Cause: {2}",
+          error.getCode(), error.getDescription(), error.getCause());
+      closed = true;
+      handleStreamClosed(error);
```

Comment:
The method has almost identical name as `handleRpcStreamClosed()`.
So `propagateStreamError()` might be a better name, as it's propagating error to the watchers.

Reply:
Hmm, I think `handleStreamClosed()` is better aligned with other "*abstract*" methods. Calling it "propagate" may not be that appropriate as this is actually how inheritance works. It is not directly calling another classes' method,  it is calling itself's method.

Reply:
Not a big deal. Just felt names are too closed.